onnxruntime/core/providers/webgpu/webgpu_context.cc

```diff
@@ -0,0 +1,677 @@
+
+        const std::unordered_map<std::string, std::string>& event_args = {
+            {"shapes", SS_GET(shapes)},
+            {"cache_key", pending_kernel_info.cache_key},
+        };
+
+        profiling::EventRecord event(profiling::API_EVENT,
+                                     -1,
+                                     -1,
+                                     pending_kernel_info.name,
+                                     static_cast<int64_t>(std::round(start_time / 1000.0)),
+                                     static_cast<int64_t>(std::round((end_time - start_time) / 1000.0)),
+                                     event_args);
+        events.emplace_back(std::move(event));
+      }
+
+      query_read_buffer.Unmap();
+      query_read_buffer.Destroy();
+    }
+
+    pending_queries_.clear();
+  }
+
+  is_profiling_ = false;
+}
+
+void WebGpuContext::EndProfiling(TimePoint /* tp */, profiling::Events& events, profiling::Events& cached_events) {
+  // This function is called when no active inference is ongoing.
+  ORT_ENFORCE(!is_profiling_, "Profiling is ongoing in an inference run.");
+
+  if (query_type_ != TimestampQueryType::None) {
+    // No pending kernels or queries should be present at this point. They should have been collected in CollectProfilingData.
+    ORT_ENFORCE(pending_kernels_.empty() && pending_queries_.empty(), "Pending kernels or queries are not empty.");
+
+    events.insert(events.end(),
+                  std::make_move_iterator(cached_events.begin()),
+                  std::make_move_iterator(cached_events.end()));
+
+    cached_events.clear();
+  } else {
+    LOGS_DEFAULT(WARNING) << "TimestampQuery is not supported in this device.";
+  }
+}
+
+void WebGpuContext::Flush() {
+  if (!current_command_encoder_) {
+    return;
+  }
+
+  EndComputePass();
+
+  if (is_profiling_ && num_pending_dispatches_ > 0) {
+    uint32_t query_count = num_pending_dispatches_ * 2;
+    current_command_encoder_.ResolveQuerySet(
+        query_set_,
+        0,
+        query_count,
+        query_resolve_buffer_,
+        0);
+
+    wgpu::BufferDescriptor bufferDescriptor;
+    bufferDescriptor.size = query_count * sizeof(uint64_t);
+    bufferDescriptor.usage = wgpu::BufferUsage::MapRead | wgpu::BufferUsage::CopyDst;
+    wgpu::Buffer query_read_buffer = device_.CreateBuffer(&bufferDescriptor);
+
+    current_command_encoder_.CopyBufferToBuffer(
+        query_resolve_buffer_,
+        0,
+        query_read_buffer,
+        0,
+        query_count * sizeof(uint64_t));
+
+    pending_queries_.emplace_back(std::move(pending_kernels_), query_read_buffer);
+    pending_kernels_.clear();
+  }
+
+  auto command_buffer = current_command_encoder_.Finish();
+  Device().GetQueue().Submit(1, &command_buffer);
+  BufferManager().RefreshPendingBuffers();
+  current_command_encoder_ = nullptr;
+  num_pending_dispatches_ = 0;
+}
+
+std::unordered_map<int32_t, std::unique_ptr<WebGpuContext>> WebGpuContextFactory::contexts_;
```

Comment:
So, this is a global variable? 
When WebGpuContext get destroyed, does  it need to access any Dawn functions? What if Dawn is already destroyed?

Reply:
There are 2 problems to be solved:

1. Dawn can be destroyed any time, and we are only able to get a notification from a callback. We didn't handle this onDeviceLost callback correctly to make sure both no resource leak and no invalid resource access.

2. if Dawn is not destroyed unexpectedly, it will be destroyed by the end of the process. There are potential issue with the order of destructing.

Reply:
The second one is the reason why we see “WebGPU device lost (2): Device was destroyed.” at the end.

Reply:
Wrap contexts_ as a function local static variable. Then at least it will get deleted before most global variables are cleared out.  


Or doing a hack. Clear out the context when deleting OrtEnv. 

```diff
--- a/onnxruntime/core/providers/webgpu/webgpu_context.cc
+++ b/onnxruntime/core/providers/webgpu/webgpu_context.cc
@@@ -20,13 -20,12 +20,16 @@@

  namespace onnxruntime {
  namespace webgpu {
++    void CleanupWebGpuContextFactory() {
++  WebGpuContextFactory::Cleanup();
++}
```

```diff
+++ b/onnxruntime/core/providers/webgpu/webgpu_context.h

@@ -26,6 +27,8 @@ class WebGpuContext;
 class ComputeContext;
 class ProgramBase;

+void CleanupWebGpuContextFactory();
+
 class WebGpuContextFactory {
  public:
   static WebGpuContext& CreateContext(int context_id,
@@ -35,6 +38,10 @@ class WebGpuContextFactory {
                                       ValidationMode validation_mode);
   static WebGpuContext& GetContext(int context_id);

+  static void Cleanup() {
+    contexts_.clear();
+  }
```

```diff
--- a/onnxruntime/core/session/ort_env.cc
+++ b/onnxruntime/core/session/ort_env.cc
@@ -25,7 +25,15 @@ OrtEnv::OrtEnv(std::unique_ptr<onnxruntime::Environment> value1)
     : value_(std::move(value1)) {
 }

+namespace onnxruntime {
+
+namespace webgpu {
+
+void CleanupWebGpuContextFactory();
+}
+}  // namespace onnxruntime
 OrtEnv::~OrtEnv() {
+  onnxruntime::webgpu::CleanupWebGpuContextFactory();
 // We don't support any shared providers in the minimal build yet
 #if !defined(ORT_MINIMAL_BUILD)
   UnloadSharedProviders();
```